src/StaticLS/IDE/CodeActions/RemoveRedundantImports.hs

```diff
@@ -0,0 +1,185 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards #-}
+
+module StaticLS.IDE.CodeActions.RemoveRedundantImports where
+
+import AST qualified
+import AST.Haskell qualified as Haskell
+import Control.Applicative
+import Control.Exception
+import Control.Monad
+import Control.Monad.IO.Class
+import Control.Monad.Trans.Maybe
+import Data.Edit
+import Data.Foldable
+import Data.LineCol
+import Data.List qualified as List
+import Data.Map ()
+import Data.Maybe
+import Data.Path
+import Data.Pos (Pos (..))
+import Data.Range (Range (..))
+import Data.Range qualified as Range
+import Data.Rope (Rope)
+import Data.Rope qualified as Rope
+import Data.Text qualified as Text
+import Data.Text.IO qualified as TextIO
+import StaticLS.IDE.CodeActions.Types
+import StaticLS.IDE.Diagnostics
+import StaticLS.IDE.Diagnostics.ParseGHC
+import StaticLS.IDE.FileWith
+import StaticLS.IDE.Monad as Monad
+import StaticLS.IDE.SourceEdit as SourceEdit
+import StaticLS.Monad
+import StaticLS.Semantic
+import StaticLS.StaticEnv
+import StaticLS.Tree
+import System.FilePath
+
+ghcidFile :: FilePath
+ghcidFile = "ghcid.txt"
+
+codeAction :: CodeActionContext -> StaticLsM [Assist]
+codeAction CodeActionContext {path, lineCol, pos} = do
+  diagnostics <- getDiagnostics
+  let relevantDiagnostics = filter isRedundantImportDiagnostic diagnostics
+  deletionInfos <- traverse mkDeletionInfo relevantDiagnostics
+  globalRedundantImportDiagnostics <- filterM (\case Full f -> isFileSaved f.path; Partial _ -> pure False) deletionInfos
+  let localRedundantImportDiagnostics = filter (isDeletionInFile path) deletionInfos
+  let globalAssist = deletionsToAssist "Remove redundant imports in project" globalRedundantImportDiagnostics
+  let localAssist = deletionsToAssist "Remove redundant imports in file" localRedundantImportDiagnostics
+  pure [localAssist, globalAssist]
+
+getDiagnostics :: StaticLsM [Diagnostic]
+getDiagnostics = do
+  staticEnv <- getStaticEnv
+  let wsRootPath = toFilePath staticEnv.wsRoot
+  let makeAbsPath = unsafeFilePathToAbs . (wsRootPath System.FilePath.</>) . toFilePath
+  let ghcidPath = wsRootPath System.FilePath.</> ghcidFile
+  info <- liftIO $ catch @IOException (TextIO.readFile ghcidPath) (const $ pure "")
+  let diagnostics = parse makeAbsPath info
+  pure diagnostics
+
+data DeletionInfo = Partial PartialDeletionInfo | Full FullDeletionInfo
+
+data PartialDeletionInfo = PartialDeletionInfo {}
+
+data FullDeletionInfo = FullDeletionInfo
+  { path :: AbsPath
+  , sourceRope :: Rope
+  , haskell :: Haskell.Haskell
+  , loc :: Range
+  , node :: Maybe Haskell.Import
+  , isPartial :: Bool
+  }
+
+mkDeletionInfo :: Diagnostic -> StaticLsM DeletionInfo
+mkDeletionInfo diagnostic
+  | isPartialDiagnostic diagnostic = pure $ Partial PartialDeletionInfo {}
+  | otherwise = Full <$> mkFullDeletionInfo diagnostic
+
+mkFullDeletionInfo :: Diagnostic -> StaticLsM FullDeletionInfo
```

Comment:
Might be nice to use mtl so that you can mock the functionality here more easily